### PR TITLE
Block definitions

### PR DESCRIPTION
**Pull request by nickynick**
This is the implementation of an idea from #467. Please see `TyphoonBlockDefinition` for details :) There is already a bunch of tests for this - not sure if that's enough, but it's something to begin with. I'll be adding some docs shortly.

As a bonus, I also felt entitled to do some cleanup & reorganizing. In particular, all infrastructure-related stuff was moved from `Typhoon.h` umbrella header to `Typhoon+Infrastructure.h`. I feel like that's the way to go with all the stuff which is not intended for basic usage.

### CODE REVIEW

File: Tests/Factory/Assembly/TestAssemblies/TyphoonConfigAssembly.m [hunk 1]
```diff
@@ -30,4 +32,13 @@ - (id)anotherKnight
     }];
 }
 
+- (id)blockKnight
+{
+    return [TyphoonBlockDefinition withClass:[Knight class] block:^id{
```
Comment (corristo) [inline Tests/Factory/Assembly/TestAssemblies/TyphoonConfigAssembly.m]: Is duplication of `Knight` class reference intentional? I mean, it's being used there and later in `[[Knight alloc] init]`.

Comment (nickynick) [inline Tests/Factory/Assembly/TestAssemblies/TyphoonConfigAssembly.m]: Yes. We need to know the component type beforehand so Typhoon would be able to use this definition for auto-injections and injections by property type.

Comment (corristo) [inline Tests/Factory/Assembly/TestAssemblies/TyphoonConfigAssembly.m]: Thanks!

**Comment by jasperblues** [comment]
Thank you @nickynick

**Comment by etolstoy** [comment]
Congrats! A great feature!

**Comment by nickynick** [comment]
:tada: :beers:  Thanks! I'll get on docs and stuff like that shortly.

**Pull request merged**